### ISSUE TITLE
Predicated Instruction support in memval_simple

### ISSUE THREAD

**Issue by toshipiazza** [DynamoRIO/dynamorio#2479]
Working on the ARM port for `memval_simple` (XREF #1551) and, at least on my machine I see consistent crashes before `main` in the following basic block:

```
...
 +77  L3  95032730   str.ls %r2 -> -0x0730(%r3)[4byte]
...
 +84  m4  e5944000   ldr    (%r4)[4byte] -> %r4
...
```

The crash is at (+84), when we're trying to dereference `%r4`, which is computed via `drutil_insert_get_mem_addr()`, i.e. `%r3 - 0x730`. When I run through this in gdb without DR, it appears as though `%r3 - 0x730` does not point to valid memory. The native app doesn't crash though, since the predicated instruction doesn't perform the store.

A (very temporary) fix would be the following patch, which I've confirmed works with no crash:

```diff
diff --git a/api/samples/memval_simple.c b/api/samples/memval_simple.c
index 20970f9..e8622cc 100644
--- a/api/samples/memval_simple.c
+++ b/api/samples/memval_simple.c
@@ -341,6 +341,8 @@ event_app_instruction(void *drcontext, void *tag, instrlist_t *bb,
         return DR_EMIT_DEFAULT;
     if (!instr_writes_memory(instr))
         return DR_EMIT_DEFAULT;
+    if (instr_is_predicated(instr))
+        return DR_EMIT_DEFAULT;

     /* XXX: See above, in handle_post_write(). To simplify the handling of registers, we
      * assume no instruction has multiple distinct memory destination operands.
```

though I would prefer a solution which allows us to still instrument predicated instructions, say something like an app2app drmgr routine like `drutil_expand_predicated`, analogous to the `drutil_expand_rep_string` routine we use in many samples.

For the record, I've never seen a crash caused by predicated instructions on x86.

**Comment by derekbruening** [comment]
Generally clients are advised to insert instrumentation as normal for a predicated instruction but mark that instrumentation with the same predicate as the app instruction.  This is what drcachesim does today.  #1723 already covers further automating this.

**Comment by toshipiazza** [comment]
Does this approach (adding the predicate to each meta instruction) work as well for clean calls, which may be emitted in the case of `drx_buf_insert_buf_memcpy`?

Also, for the automated approach, how would we insert the right predicate in the case that we're inserting instrumentation *after* the current instruction, as in this sample? Would this be based off of `XL8` values?

**Comment by derekbruening** [comment]
The only case where predicates don't work is for conditional jumps.

Inserting after is generally not supported very well: I thought you weren't actually doing that, but instead were waiting until the next instr and inserting before it?

**Comment by toshipiazza** [comment]
Yes, we're waiting until the next instruction and inserting instrumentation in the sample. Just wanted to know if an automated analysis could figure out this was the case and use the predicate of the previous instruction as opposed to the current.

**Comment by derekbruening** [comment]
> Just wanted to know if an automated analysis could figure out this was the case and use the predicate of the previous instruction as opposed to the current.

If the solution relied on instrlist insertion routines taking the predicate of the "where" instruction it might "just work".

**Comment by toshipiazza** [comment]
Just wondering, but I remember seeing references to #1702 in `{memval,instrace}_simple` and was wondering if this could also pose a problem when porting to ARM.

Unfortunately I don't know enough about thumb mode in general; specifically I don't know what should be done in the case that we need to insert a clean call in front of a predicated instruction, i.e. because `drx_buf_insert_buf_memcpy()` was given a non gpr-sized write. Do we need to change the implementation of the slowpath on ARM to veer away from the clean call?

**Comment by derekbruening** [comment]
#1702 should be solvable as part of dr_insert_it_instrs() but may require clean call out-of-line changes too: I put a comment there.
